pkg/engine/resourcecache/externalapi/externalapi.go

```diff
@@ -0,0 +1,192 @@
+package externalapi
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	kyvernov1 "github.com/kyverno/kyverno/api/kyverno/v1"
+	"github.com/kyverno/kyverno/api/kyverno/v2alpha1"
+	"github.com/kyverno/kyverno/pkg/engine/apicall"
+	"github.com/kyverno/kyverno/pkg/engine/resourcecache/cache"
+)
+
+type Poller interface {
+	Run(context.Context, <-chan struct{})
+}
+
+type Getter interface {
+	Get() (interface{}, error)
+	Stop()
+}
+
+type ExternalAPILoader struct {
+	logger logr.Logger
+	config apicall.APICallConfiguration
+	cache  cache.Cache
+}
+
+type externalEntry struct {
+	sync.Mutex
+	logger    logr.Logger
+	call      *kyvernov1.APICall
+	ticker    *time.Ticker
+	apicaller *apicall.APICall
+	data      interface{}
+	cancel    context.CancelFunc
+}
+
+func (e *externalEntry) Getter() Getter {
+	return e
+}
+
+func (e *externalEntry) Poller() Poller {
+	return e
+}
+
+func (e *externalEntry) Get() (interface{}, error) {
+	e.Lock()
+	defer e.Unlock()
+	e.logger.V(6).Info("cache entry data", "data", e.data)
+	return e.data, nil
+}
+
+func (e *externalEntry) Stop() {
+	e.cancel()
+}
+
+func (e *externalEntry) Run(ctx context.Context, stopCh <-chan struct{}) {
+	go func() {
+		for {
+			select {
+			case <-e.ticker.C:
+				data, err := e.apicaller.Execute(ctx, e.call)
+				if err != nil {
+					e.logger.Error(err, "failed to get data from api caller")
+					return
+				}
+				e.Lock()
```

Comment:
Same here. You can use `RLock()`.

Reply:
We should add `rlocks` only to reads right? I am writing a value there

Reply:
RLocks locks writing to one goroutine and allows any one to read. Locks locks writing and reading to only one goroutine at a time.

Reply:
I mean, we have to use Lock() here as we are writing. even after we move to RWMutex. as Rlock is the read lock and Lock is the write lock for RW